slapo/schedule.py

```diff
@@ -341,6 +329,8 @@ def hook_func(_module, _input, output):
 
         if mode in {"backward", "both"}:
             # Case 1, 2
+            if sync_op is not None:
+                raise ValueError("sync_op is not supported in backward yet")
```

Comment:
i am a bit confused why sync_op is not supported for backward. Don't we need corresponding op for reduce scatter and allgather?

Reply:
See the implementation of all_gather_forward_output.

Reply:
i see. so they have been coupled. then, why don't we use the same rule for allreduce since backward behavior is always coupled with forward.

Reply:
Because it's not always coupled.

Case 1: Partitioned output with all-gather.
```
Forward: all-gather
Backward: split -> allreduce.
```
In this case, all-gather and split are coupled, so we use `all_gather_forward_output`, but backward needs a standalone allreduce. We use `sch.sync(mode="both")`.

Case 2: Partitioned output without all-gather.
```
Forward: N/A
Backward: allreduce.
```
In this case, backend needs a standalone allreduce, so we use `sch.sync(mode="backward")`.

Case 3: Partial sum with allreduce.
```
Forward: allreduce
Backward: Identical
```
In this case, backward doesn't need anything, so we just `sch.sync(mode="forward")`.